File: MarsRover/python/mars_rover.py
"""Mars Rover — terminal tunnel game.

A rover sits at the bottom-centre of the screen.
A tunnel with random curves scrolls downward.
Steer left/right to avoid the walls.
"""

import sys
import os
import time
import random
import tty
import termios
import select

# ── Configuration ────────────────────────────────────────────────────
WIDTH = 60
HEIGHT = 25
TUNNEL_WIDTH = 14
ROVER_CHAR = "A"
WALL_CHAR = "█"
EMPTY_CHAR = " "
DELAY = 0.07  # seconds between frames


def init_tunnel(width, tunnel_w):
    """Return initial left-wall position centred on the screen."""
    return (width - tunnel_w) // 2


def get_key():
    """Non-blocking key read (Unix)."""
    if select.select([sys.stdin], [], [], 0)[0]:
        return sys.stdin.read(1)
    return None


def clear_screen():
    sys.stdout.write("\033[2J\033[H")


def main():
    rows = HEIGHT
    cols = WIDTH
    tunnel_w = TUNNEL_WIDTH
    left = init_tunnel(cols, tunnel_w)
    rover_x = cols // 2
    score = 0

    # Build initial tunnel map (list of left-wall positions)
    tunnel = []
    for _ in range(rows):
        tunnel.append(left)
        drift = random.choice([-1, 0, 0, 1])
        left = max(1, min(cols - tunnel_w - 1, left + drift))

    # Set terminal to raw mode for non-blocking input
    old_settings = termios.tcgetattr(sys.stdin)
    try:
        tty.setcbreak(sys.stdin.fileno())

        alive = True
        while alive:
            # ── Input ─────────────────────────────────────────────
            key = get_key()
            if key == "q" or key == "Q":
                break
            elif key in ("\x1b",):  # escape sequence
                k2 = get_key()
                if k2 == "[":
                    k3 = get_key()
                    if k3 == "D":        # left arrow
                        rover_x = max(0, rover_x - 1)
                    elif k3 == "C":      # right arrow
                        rover_x = min(cols - 1, rover_x + 1)
            elif key == "a" or key == "A":
                rover_x = max(0, rover_x - 1)
            elif key == "d" or key == "D":
                rover_x = min(cols - 1, rover_x + 1)

            # ── Scroll tunnel (new row at top, scrolls toward rover) ──
            tunnel.pop()
            drift = random.choice([-1, -1, 0, 0, 0, 1, 1])
            left = tunnel[0] + drift
            left = max(1, min(cols - tunnel_w - 1, left))
            tunnel.insert(0, left)

            # ── Collision check ───────────────────────────────────
            rover_row = rows - 2  # rover is on second-to-last row
            tl = tunnel[rover_row]
            tr = tl + tunnel_w
            if rover_x <= tl or rover_x >= tr:
                alive = False

            score += 1

            # ── Render ────────────────────────────────────────────
            buf = []
            buf.append("\033[2J\033[H")
            buf.append(f"  MARS ROVER  |  Score: {score}\n")
            for row in range(rows):
                tl = tunnel[row]
                tr = tl + tunnel_w
                line = []
                for c in range(cols):
                    if row == rover_row and c == rover_x:
                        line.append(ROVER_CHAR)
                    elif c <= tl or c >= tr:
                        line.append(WALL_CHAR)
                    else:
                        line.append(EMPTY_CHAR)
                buf.append("".join(line))
                buf.append("\n")
            buf.append("  Arrow keys / A,D to steer  |  Q to quit\n")
            sys.stdout.write("".join(buf))
            sys.stdout.flush()

            time.sleep(DELAY)

        # ── Game over ─────────────────────────────────────────────
        clear_screen()
        print(f"\n  GAME OVER!  Final Score: {score}\n")

    finally:
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, old_settings)


if __name__ == "__main__":
    main()

File: MarsRover/javascript/mars_rover.js
/**
 * Mars Rover — JavaScript (Node.js)
 *
 * Tunnel-running game. Rover at the bottom, tunnel scrolls down.
 * Steer left/right to avoid walls.
 */

const WIDTH    = 60;
const HEIGHT   = 25;
const TUNNEL_W = 14;
const DELAY_MS = 70;

let tunnel = [];
let roverX = Math.floor(WIDTH / 2);
const roverRow = HEIGHT - 2;
let score = 0;
let alive = true;

// Init tunnel
let left = Math.floor((WIDTH - TUNNEL_W) / 2);
for (let i = 0; i < HEIGHT; i++) {
    tunnel.push(left);
    left += Math.floor(Math.random() * 3) - 1;
    left = Math.max(1, Math.min(WIDTH - TUNNEL_W - 1, left));
}

// Raw mode for keyboard input
process.stdin.setRawMode(true);
process.stdin.resume();
process.stdin.setEncoding('utf8');

process.stdin.on('data', (key) => {
    if (key === 'q' || key === 'Q' || key === '\u0003') {
        alive = false;
        return;
    }
    if (key === '\u001b[D' || key === 'a' || key === 'A') { // left
        if (roverX > 0) roverX--;
    }
    if (key === '\u001b[C' || key === 'd' || key === 'D') { // right
        if (roverX < WIDTH - 1) roverX++;
    }
});

function scrollTunnel() {
    // New row at top, scrolls toward rover
    tunnel.pop();
    let drift = Math.floor(Math.random() * 5) - 2;
    let nl = tunnel[0] + drift;
    nl = Math.max(1, Math.min(WIDTH - TUNNEL_W - 1, nl));
    tunnel.unshift(nl);
}

function render() {
    const lines = ['\x1B[2J\x1B[H'];
    lines.push(`  MARS ROVER  |  Score: ${score}`);
    for (let r = 0; r < HEIGHT; r++) {
        const tl = tunnel[r];
        const tr = tl + TUNNEL_W;
        let row = '';
        for (let c = 0; c < WIDTH; c++) {
            if (r === roverRow && c === roverX) row += 'A';
            else if (c <= tl || c >= tr) row += '#';
            else row += ' ';
        }
        lines.push(row);
    }
    lines.push('  Arrow keys / A,D to steer  |  Q to quit');
    process.stdout.write(lines.join('\n') + '\n');
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function main() {
    while (alive) {
        scrollTunnel();
        score++;

        // Collision
        const tl = tunnel[roverRow];
        const tr = tl + TUNNEL_W;
        if (roverX <= tl || roverX >= tr) {
            alive = false;
        }

        render();
        if (!alive) break;
        await sleep(DELAY_MS);
    }

    process.stdout.write(`\x1B[2J\x1B[H\n  GAME OVER!  Final Score: ${score}\n\n`);
    process.stdin.setRawMode(false);
    process.exit(0);
}

main();
